Modules/_zoneinfo.c

```diff
@@ -2205,7 +2205,15 @@ find_ttinfo(zoneinfo_state *state, PyZoneInfo_ZoneInfo *self, PyObject *dt)
     }
 
     unsigned char fold = PyDateTime_DATE_GET_FOLD(dt);
-    assert(fold < 2);
+
+    // gh-125318: out-of-bounds sanity check on non-PyDateTime types
+    if (fold >= 2) {
+        PyErr_Format(PyExc_MemoryError,
```

Comment:
Don't use a `MemoryError`:
```suggestion
        PyErr_Format(PyExc_ValueError,
```